File: LeetCode/286-walls-and-gates/ex_286.cpp
class Solution {
public:
    void wallsAndGates(vector<vector<int>>& rooms) {
        for (int row = 0; row != rooms.size(); ++row) {
            for (int col = 0; col != rooms[0].size(); ++col) {
                if (rooms[row][col] == 0)
                    dfs(rooms, row, col, 0);
            }
        }
    }
private:
    vector<pair<int, int>> offset{{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
    void dfs(vector<vector<int>>& rooms, int row, int col, int distance) {
        rooms[row][col] = min(distance, rooms[row][col]);
        for (auto& p : offset) {
            int next_row = row + p.first;
            int next_col = col + p.second;
            if (next_row >= 0 && next_row < rooms.size() && next_col >= 0 && next_col < rooms[0].size() && rooms[next_row][next_col] > rooms[row][col])
                dfs(rooms, next_row, next_col, distance + 1);
        }
    }
};

File: LeetCode/286-walls-and-gates/ex_286.py
class Solution(object):
    def wallsAndGates(self, rooms):
        """
        :type rooms: List[List[int]]
        :rtype: void Do not return anything, modify rooms in-place instead.
        """
        if not rooms:   return
        for row in xrange(len(rooms)):
            for col in xrange(len(rooms[0])):
                if rooms[row][col] == 0:
                    self.dfs(rooms, row, col, 0)
    
    def dfs(self, rooms, row, col, distance):
        rooms[row][col] = min(distance, rooms[row][col])
        shift = [(1, 0), (-1, 0), (0, 1), (0, -1)]
        for i, j in shift:
            next_row = row + i
            next_col = col + j
            if 0 <= next_row < len(rooms) and 0 <= next_col < len(rooms[0]) and rooms[next_row][next_col] > rooms[row][col]:
                self.dfs(rooms, next_row, next_col, distance + 1)